rfcs/text/0011_persistable_state_service.md

```diff
@@ -0,0 +1,218 @@
+
+We also need to make sure that all the persited state containing references to saved objects extracts those before being saved and injects them later.
+
+# Detailed design
+
+```ts
+interface PersistableState extends Serializable {}
+
+interface PersistableStateDefinition<P extends PersistableState = PersistableState> {
+  id: string,
+  // migrate function receives state and version string and should return latest state version
+  // default is identity function
+  migrate: (state: unknown, version: string) => P,
+  // inject function receives state and a list of references and should return state with references injected
+  // default is identity function
+  inject: (state: P, references: SavedObjectReference[]) => P,
+  // extract function receives state and should return state with references extracted and array of references
+  // default returns same state with empty reference array
+  extract: (state: P) => { state: P, references: SavedObjectReference[] }
+}
+
+class PersistableStatePlugin {
+  setup(core, plugins) {
+    return {
+      // if one of the functions is not provided default is used 
+      // so we can always assume that all of the functions are available for every persistable state 
+      register: (id: string, definition: Partial<PersistableStateDefinition>) => void
+    }
+  }
+  start(core, plugins) {
+    return {
+      // get will always return. If requested persistable state is not defined it will return default set of functions.
+      get: (id: string) => PersistableStateDefinition,
+      // takes the state, extracts the references and returns them + version string
+      beforeSave: (id: string, state: PersistableState) => [PersistableState, SavedObjectReference[], string],
+      // takes the state, references and version and returns latest (migrated) state with references injected
+      afterLoad: (id: string, state: PersistableState, references: SavedObjectReference[], version: string) => PersistableState,
+    }
+  }
+}
+```
+
+
+# Consuming and edge cases
+
+When plugin A wants to store some state that it does not own it should check with persistableStateRegistry for a registered persistable state definition and execute `beforeSave` function.
+
+```ts
+  const stateForSaving = persistableStatePlugin.beforeSave(id, myState);
+
+``` 
+
+WARNING: If state id is known at compile time we should rather import the correct utilities directly from expression plugin. As using the registry can hide dependencies between plugins when possible plugins should directly depend on plugins providing the state they want to store.
+
+## Handling corrupt state
+
+In current proposal handling of corrupt state is up to the implementator of PersistableStateDefinition. State incoming to migrate function could not match with version string provided so author should validate the incoming state first.
+
+## EnhacedDrilldownEmbeddableInput
```

Comment:
Created enhancements issue because it didn't exist before and I never checked my PR in - https://github.com/elastic/kibana/issues/68880.  There were some unresolved comments and I moved on to other things.

As for review, I think we should email out to Kibana when it's ready for external feedback, then consumers of this service (outside app arch team) will be aware and can weigh in.  I suspect most Kibana devs won't know they need to be aware of it, so rather than headline it something like "Persistable state registries RFC" where no one knows what that means, you could frame it like "How to ensure your code doesn't break when public APIs change".  It's more complicated than that, but if we frame it simply, we'll likely get more eyes on the RFC.

A link to one of the presentations in the introduction of this RFC might be useful to provide this background context, as well as embedding one of the images that shows the different roles:  Registrator, Registrar, Persister and Enhancer.

Reply:
not sure if we can link to representations as they are on gdrive and this is gonna be in public repo ?